### ISSUE TITLE
Async data load without UI blocking

### ISSUE THREAD

**Issue by rokkolesa** [vaadin/flow-components#1066]
### Description
When adding a `Grid` to a view with an `AbstractBackEndDataProvider` added, the view takes a longer time to display. This is due to it retrieving the data from the `DataProvider`, which can take a long time depending on the data source. All of this is done synchronously blocking the whole rendering of the view.

What I would like is to display the `Grid` with an overlay of some sort displaying a message (e.g. `Please wait, the data is loading`) with no rows, adding the rows when the actual load finishes and removing the message.

I tried to get around this by telling the `DataProvider` to not load any data initially - like "disabling" the `DataProvider` data fetch by returning a count of 0 and an empty `Stream` of data. Then I delegated the fetch to the next request using something like this:
```java
getElement().executeJs("")
            .then(json -> {
                <enable DataProvider load >
                grid.getDataProvider().refreshAll();
                <close the messafge before client response>
            });
```
The results were okay, the data was loaded only after the view rendered. However, the UI was blocked when the load was in progress and the user had to wait for it to finish.

I did some digging and found that the `DataCommunicator` always adds a function for execution before client response of the current request on a `DataChangeEvent` (method `requestFlush`.. it also adds it on initial attach). This means that there is no way of making the data load truly async as the UI is always blocked when this event is triggered otherwise an exception is thrown that the session is not locked.

Ideally the UI would not be blocked and the user would be able to interact with other components on the view as well as being able to navigate away from the view if the data load would take a long time, cancelling the data fetch. Is there a way to do this in the current implementation?

### Expected outcome
The data load should not block the view rendering.

### Actual outcome
Data load blocks the view rendering, not showing any changes until the data load is finished.

### Browsers Affected
All

### Versions:
- Vaadin / Flow / Grid version: `14.4.3` / `2.4.2` / `4.3.3`
- Java version: `11`
- OS version: `macOS Big Sur 11.0.1`
- Browser version (if applicable): `Chrome 87.0.4280.67`
- Application Server (if applicable): `WildFly 21.0.0`

**Comment by TatuLund** [comment]
I made proof of concept design with Vaadin 14. This can be quite nicely dealt in DataCommunicator and no changes are needed in Grid or DataProvider https://gist.github.com/TatuLund/a97e01702705d2a53a67c114e514d8cd

**Comment by lanmaster** [comment]
[Another issue like this](https://github.com/vaadin/flow/issues/12342)

**Comment by TatuLund** [comment]
Also https://github.com/vaadin/flow/issues/10709
